### PR TITLE
Scout Package and Repo Reorganization

### PR DESCRIPTION
**Pull request by aspeake**
This PR reorganizes the repo to define a `scout` package. This will allow for better reproducibility across machines, better organization and therefore more efficient development, and potential for re-use outside of the repo. A summary of the new structure for the repo is as follows:
- The `./scout` package contains all modules (eg, ecm_prep.py)
- `./tests` directory contains all files related to tests
- **_Package data_** is saved to scout/supporting_data and installed as part of the package
- **_Non-package data_** (`./ecm_definitions`, `./results`, `./inputs`, `./generated` files) are saved to the parent directory and not installed with the package
- A new directory `./generated` is created to store intermediate files (such as ecm_prep.json)
- Specific files not categorized as modules or tests were put in either the `./inputs` directory or an appropriate folder within `./scout/supporting_data`

image

With changes to filepaths, steps were taken to generalize how paths are defined for reading and writing throughout the repo. Specifically, a new class `scout.constants.FilePaths` was created to house the most commonly used paths, and that is now incorporated in many of the modules using the `pathlib` package. This results in cleaner code in many instances and a more general way to make updates.

### CODE REVIEW

File: scout/ecm_prep.py [hunk 1]
```diff
@@ -1478,10 +1415,10 @@ def __init__(self, base_dir, handyfiles, opts):
                     metrics_data = handyfiles.tsv_metrics_data_net_hr
 
                 # Import system max/min and peak/take hour load by EMM region
+                # raise ValueError(metrics_data, opts.tsv_metrics[-2])
```
Comment (aspeake) [inline scout/ecm_prep.py:1418]: remove

**Comment by aspeake** [comment]
@trynthink pulled in the latest master updates and confirmed that the full workflow for input file updates works on this branch (after a couple of fixes).

**Review by trynthink — COMMENTED** [review]
A couple broad questions:
1. Is there any value in renaming the test files all test_*.py instead of *_test.py? (It seems like it would just avoid the need for the `-p` flag with `python3 -m unittest discover`, but maybe there are other benefits I'm not aware of?
2. Once revisions are complete, are all the commits in this PR needed? Do you want to rebase and squash/fixup some of them to group changes together logically or collapse everything into a single commit?
3. In this PR or a separate set of changes, should we segregate the modules used to update input data (e.g., cambium_updater.py, com_mseg.py) and the modules used to do an analysis (ecm_prep.py and run.py)?

File: docs/installation_guide.rst [hunk 2]
```diff
@@ -25,7 +27,7 @@ If you're comfortable at the command line, install or set up everything in this
 **Prerequisites**
 
 * Python 3
-* Python packages: pip [#]_, matplotlib, numpy, numpy_financial, pandas, xlsxwriter
```
[the remark this answers is not in the document]
Comment (aspeake) [reply docs/installation_guide.rst:28]: Updated references to this footnote in two places (one for Windows, one for MacOS)

File: docs/installation_guide.rst [hunk 3]
```diff
@@ -63,22 +65,19 @@ In a Terminal window, at the command prompt (a line terminated with a $ characte
 
    brew install python3
 
-2. Install required Python packages
+2. Install Scout Python package
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-Once Python 3 is fully installed, pip3 is the tool you will use to install add-ons specific to Python 3. Several packages are required for Scout: matplotlib, numpy, numpy_financial, pandas, and xlsxwriter. To install them, at the command prompt in Terminal, type::
```
[the remark this answers is not in the document]
Comment (aspeake) [reply docs/installation_guide.rst:69]: Good catch, I re-inserted part of it to match the Windows version.

File: docs/tutorials.rst [hunk 4]
```diff
@@ -1270,16 +1270,16 @@ To run the pre-processing script |html-filepath| ecm_prep.py\ |html-fp-end|, ope
 **Windows** ::
 
    cd Documents\projects\scout-run_scheme
-   py -3 ecm_prep.py
+   py -3 ./scout/ecm_prep.py
```
[the remark this answers is not in the document]
Comment (aspeake) [reply docs/tutorials.rst]: Nope, removed throughout all the code snippets

File: scout/constants.py [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    # Non-package data:
+    _parent_dir = Path.cwd()  # parent dir of repo
+    ECM_DEF = _parent_dir / "ecm_definitions"
+    GENERATED = _parent_dir / "generated"
+    ECM_COMP = GENERATED / "ecm_competition_data"
+    EFF_FS_SPLIT = GENERATED / "eff_fs_splt_data"
+    INPUTS = _parent_dir / "inputs"
+    RESULTS = _parent_dir / "results"
+    PLOTS = RESULTS / "plots"
+    METADATA_PATH = INPUTS / "metadata.json"
```
[the remark this answers is not in the document]
Comment (aspeake) [reply scout/constants.py:22]: Not really... mostly done because it seemed to be a commonly used path across several files,

**Comment by aspeake** [comment]
> A couple broad questions:
> 
>     1. Is there any value in renaming the test files all test_*.py instead of *_test.py? (It seems like it would just avoid the need for the `-p` flag with `python3 -m unittest discover`, but maybe there are other benefits I'm not aware of?
> 
>     2. Once revisions are complete, are all the commits in this PR needed? Do you want to rebase and squash/fixup some of them to group changes together logically or collapse everything into a single commit?
> 
>     3. In this PR or a separate set of changes, should we segregate the modules used to update input data (e.g., cambium_updater.py, com_mseg.py) and the modules used to do an analysis (ecm_prep.py and run.py)?

We have discussed each of these, the only action planned at the moment is to implement submodules (3), for which I made an issue to be addressed in a future PR: https://github.com/trynthink/scout/issues/364

**Comment by aspeake** [comment]
@trynthink fyi - I have updated to using setuptools_scm (https://github.com/trynthink/scout/pull/349/commits/b11c27ebaf276b14cdf835d455f6bf80a550e0ca), which replaces the need for MANFIEST.in and enables automatic versioning. Instead of requiring manual update of versions, the version is extracted from GitHub tags and inferred to a new version, if applicable.

For example, the current version of this branch will have the following version:
image
Where setuptools_scm inferred v0.10 by incrementing the most recent tag, and there have since been 27 revisions since a hypothetical v0.10 was started. If there have been no revisions or local changes, just the tag would be output.

**Review by trynthink — APPROVED** [review]

**Pull request merged**